Add test to confirm fix for issue with log rotation and ignore older …

pghalliday opened the pull request:
…when there have been no changes

ruflin commented on filebeat/tests/system/test_prospector.py:
```diff
@@ -174,3 +174,39 @@ def test_exclude_files(self):
+        with open(testfile, 'a') as file:
+            file.write("Line {}\n".format(lines))
+        # wait for log to be read
+        time.sleep(0.2)
```
Instead of using sleep which is not necessarly reliable based on the load of the test system, I recommend to use `self.output_has(lines=1)` as you can see here: https://github.com/pghalliday/beats/blob/596/filebeat/tests/system/test_prospector.py#L97

ruflin commented on filebeat/tests/system/test_prospector.py:
```diff
@@ -174,3 +174,39 @@ def test_exclude_files(self):
+        testfile = self.working_dir + "/log/test.log"
+
+        proc = self.start_filebeat(debug_selectors=['*'])
+        time.sleep(1)
```
To guarantee filebeat started as expected, you can use `self.log_contains(...)` as you can find here: https://github.com/pghalliday/beats/blob/596/filebeat/tests/system/test_prospector.py#L39

ruflin commented on filebeat/tests/system/test_prospector.py:
```diff
@@ -174,3 +174,39 @@ def test_exclude_files(self):
+        os.rename(testfile, testfile + ".1")
+        os.mknod(testfile)
+        # wait for ignore older
+        time.sleep(2)
```
Best is probably to also use `log_contains` here and wait for the ignore_older message.

ruflin commented:
@pghalliday I added some comments to make this test more reliable. We had issues in the past with flaky tests. 

There seems to be an issue with mknod the windows and mac os x. Windows test for example: https://ci.appveyor.com/project/elastic-beats/beats/build/451/job/bddp6d6rb7qtg7io#L92

```
Traceback (most recent call last):
  File "C:\gopath\src\github.com\elastic\beats\filebeat\tests\system\test_prospector.py", line 200, in test_rotating_ignore_older_low_write_rate
    os.mknod(testfile)
AttributeError: 'module' object has no attribute 'mknod'
```

Is mknod required or could it be written differently?

pghalliday commented:
I think it's required to make an empty file on rotation to reproduce the issue but I will double check. Suggested changes in progress

ruflin commented:
@pghalliday :+1: Let me know when the changes are pushed as Github does not provide notifications for new commits.

pghalliday commented:
Ok so I updated the test and it still passes on master. However in this form I have to do some extra changes to run the test on v1.0.1 as the logging changed:

Swap this line

```
                "Closing file: {}\n".format(os.path.abspath(testfile))),
```

for

```
              "Error: Stop harvesting as file is older then ignore_older: {};".format(os.path.abspath(testfile))),
```

This threw up something interesting - this test passes on v1.0.1! So I added an additional 2 second wait after that log line and the test failed again. So I checked the logs and noticed that I get 2 loglines when I think I should only get 1:

```
11:36 $ cat build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/filebeat.log | grep ERR
2015/12/31 10:35:12.434045 log.go:102: ERR File reading error. Stopping harvester. Error: Stop harvesting as file is older then ignore_older: /Users/pghalliday/projects/go/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log; Last change was: 1.020251124s
2015/12/31 10:35:12.748363 log.go:102: ERR File reading error. Stopping harvester. Error: Stop harvesting as file is older then ignore_older: /Users/pghalliday/projects/go/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log; Last change was: 1.031843635s
```

Note that these are not from consecutive scans, there were a few in between (I checked the full log). I suspect that one of these lines should be for the `test.log.1` file (there was no error reported for that file). Maybe this is a clue to a problem with the rename functionality.

Unfortunately it may mean that this test is not so useful for regression after all and a new test might be needed to validate the rename stuff

pghalliday commented:
Just double checked the master branch with an additional 2 second wait and it still passes. However there is still some weirdness in the logs

```
12:13 $ cat build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/filebeat.log | grep Closing
2015/12/31 11:13:02.132956 log.go:116: DBG  Closing file: /Users/pghalliday/projects/go/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log
2015/12/31 11:13:02.344091 log.go:116: DBG  Closing file: /Users/pghalliday/projects/go/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log
2015/12/31 11:13:02.449277 log.go:116: DBG  Closing file: /Users/pghalliday/projects/go/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log.1
```

This time I do get a separate `Closing file:` line for `test.log.1` but I still get 2 for `test.log`

pghalliday commented:
Oh and I just checked without creating an empty file on v1.0.1 and it still fails - I thought this wouldn't work as I would not get an error for the file as it does not exist - but I do! In fact I still get 2 - no idea what's going on

Anyway it still fails waiting for the second test log line

I created a separate branch for v1.0.1 tests as the logging is different

https://github.com/pghalliday/beats/tree/1.0.1-ignore_older-with-file-rename

ruflin commented:
@pghalliday Thanks for all the updates and details. I will look more in detail into it later today.

pghalliday commented:
Updated test with additional timeout that would cause the test to fail on v1.0.1

ruflin commented:
Thanks. Merged. Thanks again for all the help. I will close https://github.com/elastic/beats/pull/604

pghalliday commented:
your welcome :)
